futurewei-cloud/arktos
# kubelet prepares pod DNS config with network specific DNS service IP

**Do not merge yet**
it requires the feature that kubelet has permission to get network CR, tracked by #447

**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
this PR changes pod DNS config (/etc/resolv.conf) - for pod of ClusterFirst DNS Policy, its name server shall be the DNS service IP address of the associated network. 

Based on [multi-tenancy network design](https://github.com/futurewei-cloud/arktos/blob/master/docs/design-proposals/multi-tenancy/multi-tenancy-network.md), pod resides in a certain network; as each network has its own DNS service, pods inside of the network should point to this network specific DNS IP. This is a breaking change - k8s uses cluster-shared DNS ip address for all pods, Arktos assigns pod with its associated network DNS address.

Whenever kubelet needs to get network CR, it sends get request to API server. This approach is simple, but not most efficient. This PR chooses this approach for its simplicity. One direction of improvement is to introduce cache to avoid unnecessary queries. This optimization can be considered in the future.

In this PR,  if the pod does not explicitly associate to a network and the default network not created yet, pod DNS config still uses  the cluster-shared DNS IP, as current k8s does. This temporary measure is not to break system before all multi-tenant dependencies are in place. It should be removed when all are ready.

**How to verify at the local dev env**:
note: in this PR, __kubelet needs permission to get networks custom resource of group "arktos.futurewei.com".__ If the kubelet of your env has not such permission yet, you can temporarily run kubelet in cluster-admin context.
1. start arktos by arktos-up.sh
2. ensure network CRD defined by ```./cluster/kubectl.sh create -f ./pkg/controller/artifacts/crd-network.yaml```
3. create tenant x by ```./cluster/kubectl.sh create tenant x```
4. inside tenant x create network net-y using following yaml
```yaml
apiVersion: "arktos.futurewei.com/v1"
kind: Network
metadata:
  name: net-y
  tenant: x
spec:
  type: flat
```
5. manually simulate (e.g. through subresource status update) net-y to have dns service ip assigned, like below
```
./cluster/kubectl.sh get ne --tenant x
NAME    TYPE   VPC   PHASE   DNS
net-y   flat         Ready   1.2.3.4
```
6. create pod inside tenat x using following pod spec:
```yaml
apiVersion: v1
kind: Pod
metadata:
  name: nginx-y
  tenant: x
  labels:
    name: nginx
    arktos.futurewei.com/network: net-y
spec:
  containers:
  - name: nginx
    image: nginx
    ports:
    - containerPort: 80
```
7. verify pod-y /etc/resolv.conf having following content - 1.2.3.4 is the DNS service ip defined in net-y:
```
search default.svc.cluster.local svc.cluster.local cluster.local
nameserver 1.2.3.4
options ndots:5
```

**Does this PR introduce a user-facing change?**:
Yes
```release-note
Arktos assigns pod /etc/resolv.config with its associated network DNS address, if pod is of ClusterFirst DNSPolicy
```
